Source/Layers/InnerProductLayer.swift

```diff
@@ -139,4 +145,58 @@ public class InnerProductLayer: ForwardLayer, BackwardLayer {
             encoder.endEncoding()
         }
     }
+
+    public func update(buffer: MTLCommandBuffer, updateFunction: MTLComputePipelineState, solverParams: MTLBuffer?) {
+        if weightDiff == nil {
+            weightDiff = buffer.device.newBufferWithLength(inputSize * outputSize, options: .CPUCacheModeDefaultCache)
+            weightDiff!.label = "InnerProductWeightDiffs"
+        }
+        if biasDiff == nil {
+            biasDiff = buffer.device.newBufferWithLength(outputSize, options: .CPUCacheModeDefaultCache)
+            biasDiff!.label = "InnerProductBiasDiffs"
+        }
+
+        do {
+            var weightDims = BufferDimensions(count: UInt32(weightsBuffer.length))
+            let weightDimsBuffer = buffer.device.newBufferWithBytes(&weightDims, length: sizeof(Parameters), options: .CPUCacheModeDefaultCache)
+            weightDimsBuffer.label = "InnerProductWeightsDimensions"
+
+            let encoder = buffer.computeCommandEncoder()
+            encoder.label = "InnerProductUpdateParamsWeights"
+            encoder.setComputePipelineState(updateFunction)
+            encoder.setBuffer(weightsBuffer, offset: 0, atIndex: 0)
```

Comment:
This is very fragile to changes in the signature of updateFunction, and this will be repeated all over the place. Instead pass an object that takes care of this. Something like `update(updateAction: (MTLBuffer, MTLBuffer) -> Void) { updateAction(weightsBuffer, weightDiff) ... }`